shopfloor_delivery_shipment/services/delivery_shipment.py

```diff
@@ -593,46 +582,87 @@ def _data_for_content_to_load_from_pickings(self, shipment_advice):
             pickings_partially_loaded = loaded_lines.picking_id
             domain += [("picking_id", "in", pickings_partially_loaded.ids)]
         move_lines = self.env["stock.move.line"].search(domain)
-        return self._prepare_data_for_content(move_lines)
+        return self._prepare_data_for_content(move_lines, False)
 
     def _data_for_content_to_load_from_picking(
         self, shipment_advice, picking=None, location=None
     ):
         """Return a dictionary where keys are source locations
         and values are dictionaries listing package_levels and move_lines
         loaded or to load.
+        """
+        # Grab move lines to sort, restricted to the current delivery
+        if picking:
+            move_lines = self._find_move_lines_to_process_from_picking(
+                shipment_advice, picking
+            )
+        elif location:
+            move_lines = self._find_move_lines_from_location(shipment_advice, location)
+        group_by_sales = bool(location)
+        return self._prepare_data_for_content(move_lines, group_by_sales)
+
+    def _prepare_data_for_content(self, move_lines, group_by_sales):
+        """Returns data for package levels and/or move lines.
+
+        The data is grouped by source location with a key for package levels and
+        a key for move lines.
+        The package levels can be grouped by related sales order and if they are not
+        they will be all stored into a key whose name is only spaces.
+        The move lines are not grouped by sales.
 
         E.g:
             {
                 "SRC_LOCATION1": {
-                    "package_levels": [{PKG_LEVEL_DATA}, ...],
+                    "package_levels": {
+                       "sale 01": [{PKG_LEVEL_DATA}, ...],
+                       "sale 02": [{PKG_LEVEL_DATA}, ...],
+                       " ": [{PKG_LEVEL_DATA}, ...],
+                    }
                     "move_lines": [{MOVE_LINE_DATA}, ...],
                 },
                 "SRC_LOCATION2": {
                     ...
                 },
             }
         """
-        # Grab move lines to sort, restricted to the current delivery
-        if picking:
-            move_lines = self._find_move_lines_to_process_from_picking(
-                shipment_advice, picking
-            )
-        elif location:
-            move_lines = self._find_move_lines_from_location(shipment_advice, location)
-        return self._prepare_data_for_content(move_lines)
-
-    def _prepare_data_for_content(self, move_lines):
+        empty_group_name = " "
```

Comment:
```python
>>> json.dumps({"package_level_1": list(range(10)), False: list(range(10))})
'{"package_level_1": [0, 1, 2, 3, 4, 5, 6, 7, 8, 9], "false": [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]}'
```
Can't you use False as the empty key? I don't like `" "`

Reply:
BTW what's the main reason to do this server side?
In theory this is simply a fronted requirement and we already have grouping functions client side (eg: group by location). See in `wms_utils`
Unless there's a real benefit I would move everything frontend side.

Reply:
well, as i understand, the sorting has to consider already loaded shipments.
Is this information already available in the frontend?
Maybe it's easier to handle this from the backend.

Reply:
I agree that " " group name was not best.
About using False or None. Having different value type has key will fail when using `json.dumps` and the `sort_key` argument like in https://github.com/OCA/rest-framework/blob/e8ce61565003639f31623354b952e5aee2df02f3/rest_log/components/service.py#L26

Anyway a None value will be transformed to a "null" string when the json is being dumped. So I set the group name to "null".